demo/login.cgi

```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/python
+r"""
+Taken from https://github.com/plone/plone.session/blob/master/plone/session/tktauth.py
+The main difference is in how the "ip" is encoded in createTicket()
+"""
+
+from struct import pack
+
+import hashlib
+import hmac
+import six
+import time
+import sys
+
+
+def safe_encode(value, encoding='utf-8'):
+    """Convert unicode to the specified encoding.
+
+    copied from Products.CMFPlone.utils b/c this package does not depend on it
+    """
+    if isinstance(value, six.text_type):
+        value = value.encode(encoding)
+    return value
+
+def mod_auth_tkt_digest(secret, data1, data2):
+    # data1: addr + timestamp
+    # data2: userid+token_list+user_data
+    digest0 = hashlib.md5(data1 + secret + data2).hexdigest()
+    if not six.PY2:
+        # In Python 3 hashlib.md5(value).hexdigest() wants a bites value
+        # and returns text
+        digest0 = digest0.encode()
+    digest = hashlib.md5(digest0 + secret).hexdigest()
+    return digest
+
+def createTicket(secret, userid, tokens=(), user_data='', ip='0.0.0.0',
+                 timestamp=None, encoding='utf-8', mod_auth_tkt=False):
+    """
+    By default, use a more compatible
+    """
+    if timestamp is None:
+        timestamp = int(time.time())
+    secret = safe_encode(secret)
+    userid = safe_encode(userid)
+    tokens = [safe_encode(t) for t in tokens]
+    user_data = safe_encode(user_data)
+
+    token_list = b','.join(tokens)
+
+    # ip address is part of the format, set it to 0.0.0.0 to be ignored.
+    # pack is used to convert timestamp from an unsigned integer to 4 bytes
+    # in network byte order.
+    # Unfortunately, some older versions of Python assume that longs are always
+    # 32 bits, so we need to trucate the result in case we are on a 64-bit
+    # naive system.
+    ##data1 = inet_aton(ip)[:4] + pack('!I', timestamp) # THIS is the line that was modified to work with lighttpd mod_authn_tkt
```

Comment:
Thank you for documenting your code this way to make it easier to read.

FYI: the reason lighttpd mod_authn_tkt does this differently is that an IP address might be IPv4 or it might be IPv6.  An IPv6 address does not fit into 4 packed bytes (which is the size of an IPv4 address)

Reply:
Ah! That explains a lot :-)
Thanks for the compliment, but I did not write the Python code, it's copied from https://github.com/plone/plone.session/blob/master/plone/session/tktauth.py which I reached from this ticket https://redmine.lighttpd.net/boards/2/topics/8674
I put the link to the original in the header, IDK if that would be enough